src/Compilers/Core/Portable/Serialization/ObjectWriter.cs

```diff
@@ -697,10 +709,14 @@ private void WriteObject(object instance)
             }
             else
             {
-                var writable = instance as IObjectWritable;
+                var writable = instanceAsWritableOpt;
```

Comment:
I started with that.  but then you have two null checks in the common case.  The first for ?? and hte second to see if you have null right after that.  In the common case i would prefer just a single null check in the common case, and two in the rare case.